src/System.Net.NameResolution/src/System/Net/DNS.cs

```diff
@@ -0,0 +1,383 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Globalization;
+using System.Net.Internals;
+using System.Net.Sockets;
+using System.Threading.Tasks;
+
+namespace System.Net
+{
+    /// <devdoc>
+    ///    <para>Provides simple
+    ///       domain name resolution functionality.</para>
+    /// </devdoc>
+
+    public static class Dns
+    {
+        // Host names any longer than this automatically fail at the winsock level.
+        // If the host name is 255 chars, the last char must be a dot.
+        private const int MaxHostName = 255;
+
+        internal static IPHostEntry InternalGetHostByName(string hostName, bool includeIPv6)
+        {
+            if (Logging.On) Logging.Enter(Logging.Sockets, "DNS", "GetHostByName", hostName);
+            IPHostEntry ipHostEntry = null;
+
+            GlobalLog.Print("Dns.GetHostByName: " + hostName);
+            NameResolutionPal.EnsureSocketsAreInitialized();
+
+            if (hostName.Length > MaxHostName // If 255 chars, the last one must be a dot.
+                || hostName.Length == MaxHostName && hostName[MaxHostName - 1] != '.')
+            {
+                throw new ArgumentOutOfRangeException("hostName", SR.Format(SR.net_toolong,
+                    "hostName", MaxHostName.ToString(NumberFormatInfo.CurrentInfo)));
+            }
+
+            //
+            // IPv6 Changes: IPv6 requires the use of getaddrinfo() rather
+            //               than the traditional IPv4 gethostbyaddr() / gethostbyname().
+            //               getaddrinfo() is also protocol independant in that it will also
+            //               resolve IPv4 names / addresses. As a result, it is the preferred
+            //               resolution mechanism on platforms that support it (Windows 5.1+).
+            //               If getaddrinfo() is unsupported, IPv6 resolution does not work.
+            //
+            // Consider    : If IPv6 is disabled, we could detect IPv6 addresses
+            //               and throw an unsupported platform exception.
+            //
+            // Note        : Whilst getaddrinfo is available on WinXP+, we only
+            //               use it if IPv6 is enabled (platform is part of that
+            //               decision). This is done to minimize the number of
+            //               possible tests that are needed.
+            //
+            if (SocketProtocolSupportPal.OSSupportsIPv6 || includeIPv6)
```

Comment:
Minor optimization, but I wonder if we should switch the other of these to be ```includeIPv6 || SocketProtocolSupportPal.OSSupportsIPv6``` ? That way we don't have to do all of the initialization stuff in OSSupprtsIPv6 if includeIPv6 is false and no one actually tries to do IPv6 stuff in the app.  But maybe that's too much of an edge case to care.